File: LeetCode/valid_parenthesis.rb
# @param {String} s
# @return {Boolean}
def is_valid(s)
    return true if s.length == 0
    return false if s.length.odd?

    parens = {
    '(' => ')',
    '[' => ']',
    '{' => '}',
    }

    stack = []
    i = 0

    s.chars.each do |paren|
        stack.push paren if parens.keys.include? paren
        if parens.values.include? paren
          return false if parens.key(paren) != stack.pop
        end
    end

    stack.empty?
end

File: LeetCode/valid-parenthesis.cpp
bool isValid(string s) {
	stack<char> st;
	map<char, char> m;
	m.insert(pair<char, char>('(', ')'));
	m.insert(pair<char, char>('{', '}'));
	m.insert(pair<char, char>('[', ']'));

	for (char& c : s) {
		if (c == '(' || c == '{' || c == '[') {
			st.push(c);
		}
		else {
			if (st.empty() || c != m[st.top()]) return false;
			st.pop();
		}
	}
	return st.size() == 0;
}
